binders/kafka-binder/spring-cloud-stream-binder-kafka/src/main/java/org/springframework/cloud/stream/binder/kafka/KafkaMessageChannelBinder.java

```diff
@@ -1548,6 +1575,32 @@ public boolean isRunning() {
 			return this.running;
 		}
 
+		@Override
+		public void handleMessage(Message<?> message) {
+
+			// if we use our own partition handler to update partition count we recalculate partition
+			if (kafkaPartitionHandler != null) {
+				kafkaPartitionHandler.setPartitionCount(getKafkaTemplate().partitionsFor(this.topic).size());
+				int partitionId = kafkaPartitionHandler.determinePartition(message);
+
+				Message<?> newMessage = MessageBuilder
+					.fromMessage(message)
+					.setHeader(BinderHeaders.PARTITION_HEADER, partitionId).build();
+
+				super.handleMessage(newMessage);
+			}
+			else {
+				super.handleMessage(message);
+			}
+		}
+
+		PartitionHandler getKafkaPartitionHandler() {
+			return kafkaPartitionHandler;
+		}
```

Comment:
Do we need this accessor and setter below? I don't see where they are called from.

Reply:
I invented this getter for the test I wrote.

Reply:
Ok, I see that now. I wonder if we could find a different way to do those assertions. We don't typically add production code if it is only required by the tests, as that seems like a slight code smell. It is ok to use `ReflctionUtils` in the test to access the private fields.

Reply:
I got your point. Will try to use Reflection instead.